### PR TITLE
Sketch dist-ckpt content versioning

### PR DESCRIPTION
**Pull request by mikolajblaz**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: docs/source/checkpoints/dist_ckpt.rst [hunk 1]
```diff
@@ -574,6 +598,14 @@ FAQs
 
    To accelerate checkpoint saving, it is recommended to set ``dist_ckpt_assume_constant_structure=True``.
 
+**9. Q: I get an error about an "invalid access pattern". What does it mean?**
```
Comment (mikolajblaz) [inline docs/source/checkpoints/dist_ckpt.rst:626]: A bonus unrelated to the MR :)

File: nemo/lightning/megatron_parallel.py [hunk 2]
```diff
@@ -711,12 +711,18 @@ def sharded_state_dict(self, prefix: str = "", metadata: Optional[dict] = None)
...
 
         if metadata is None:
-            # TODO: can we assume MegatronStrategy as trainer?
-            metadata = self.trainer.sharded_state_dict_metadata
-            logging.debug(
-                f'No sharded_state_dict metadata passed for the model,'
-                f' using metadata for checkpoint save: {metadata}'
-            )
+            strategy = self.trainer.strategy
+            if isinstance(strategy, MegatronParallel):
```
Comment (ananthsub) [inline nemo/lightning/megatron_parallel.py]: should be MegatronStrategy? but megatron parallel should be used only with megatron strategy regardless

Comment (mikolajblaz) [reply nemo/lightning/megatron_parallel.py]: right, thanks, so removed the check altogether

**Comment by chtruong814** [comment]
The export test failures are unrelated to this PR.  You can get a review and we can merge if everything else is good

**Comment by mikolajblaz** [comment]
> The export test failures are unrelated to this PR. You can get a review and we can merge if everything else is good

Thanks. Added one last small change (remove `metadata['chained_optim_avoid_prefix'] = True` flag because the corresponding MCore MR is not merged yet) which restarted the pipeline, but the MR is ready. @ananthsub can you do the final review please?

**Review by ananthsub — DISMISSED** [review]

**Comment by mikolajblaz** [comment]
@chtruong814 I resolved conflicts with main and got an approval from Ananth (dimissed because of merge main). What's the next step?

**Review by dimapihtar — APPROVED** [review]
LGTM. Thank you!
